Atomic and thread operators

binji opened the pull request:
This is an initial pass at adding some atomic and thread operators, based on some work that @flagxor did earlier.

There's still a lot that needs to be fleshed out, but I wanted to get something going.

jfbastien commented on Semantics.md:
?

jfbastien commented on Semantics.md:
```diff
@@ -81,7 +81,7 @@ Note that the value types `i32` and `i64` are not inherently signed or
 operators.
 
-## Linear Memory
+## Linear memory
```
?

jfbastien commented on Semantics.md:
```diff
@@ -112,10 +112,15 @@ or [defined inside the module](Modules.md#linear-memory-section). After import
+memory if the module's memory import doesn't specify that it allows shared
+memory.
+
+### Linear memory accesses
```
Capitalization.

jfbastien commented on Semantics.md:
```diff
@@ -157,6 +162,146 @@ Store operators do not produce a value.
+| `i64.atomic_xchg32_u` | 4 bytes, zero-extended i32 to i64 | nop | 4 bytes, wrapped from i64 to i32 |
+| `i64.atomic_xchg` | 8 bytes | nop | 8 bytes |
+
+The atomic compare exchange RMW operators take three operands: a linear memory
```
Kinda weird to classify cmpxchg as RMW. It's a 3rd category IMO.

binji replied:
Done.

jfbastien commented on Semantics.md:
```diff
@@ -211,7 +356,7 @@ Thus, it is recommend that WebAssembly producers align frequently-used data to
 greatest alignment values practical, while always avoiding misaligned accesses.
 
-### Out of Bounds
+### Out of bounds
```
?

jfbastien commented on Semantics.md:
```diff
@@ -461,7 +606,7 @@ supported (including NaN values of all possible bit patterns).
   * `f64.const`: produce the value of an f64 immediate
 
-## 32-bit Integer operators
+## 32-bit integer operators
```
?

jfbastien commented on Semantics.md:
```diff
@@ -678,6 +823,48 @@ outside the range which rounds to an integer in range) traps.
+## Thread operators
+
+  * `is_lock_free`: return 1 if an atomic memory access of a given size can be
+    completed without the execution engine acquiring a lock. If not, return `0`.
```
Can you import the wording from https://tc39.github.io/ecma262/ ?

binji replied:
Done.

jfbastien commented on Semantics.md:
```diff
@@ -678,6 +823,48 @@ outside the range which rounds to an integer in range) traps.
 
+## Thread operators
+
+  * `is_lock_free`: return 1 if an atomic memory access of a given size can be
```
Code-quote `1`.

binji replied:
Done.

jfbastien commented on Semantics.md:
```diff
@@ -678,6 +823,48 @@ outside the range which rounds to an integer in range) traps.
+memory.
+
+The wait operators take three operands: a linear memory index, an
+expected value, and a relative timeout in milliseconds. The wait operation
```
Can it be negative?

binji replied:
Described below, but I moved it up.

binji commented on Semantics.md:
Capitalization was inconsistent in the file, but seemed to be "Upper lower lower." Probably should be a different change though.

binji commented:
> We need a way to identify post-MVP features. :atom_symbol: is an easy on here! We then need a page that identifies all these symbols (MVP, each post-MVP, and :unicorn: as "future").

Good point, perhaps we should open a separate issue to discuss?

> This is also missing the specification of how non-atomic accesses behave. We could just wait for the JS work to finish thought?

Yeah, I was considering just referencing the ecma262 memory model, but it's not immediately obvious to a reader how it would map.

> How do we intend to move such proposals forward?

I seem to remember there being a plan about this. Little help, anyone? :-)

lukewagner commented:
I think we're still figuring out what, post-MVP, is the process for proposing extensions.  Also, with the concurrent emergence of a proper spec in the spec repo, the shape of a PR/proposal might be quite different.

jfbastien commented:
I'm not sure we want to block everything on the spec repo at this point in time.

lukewagner commented:
I guess it depends on how long each of these things (process for evolving post-MVP, spec repo, threads) take.

flagxor commented:
Agree we don't want to block if possible.
@lukewagner would pulling out the sections this adds into a separate Threads.md doc be easier in terms of weathering the shifting repos?

AndrewScheidecker commented on BinaryEncoding.md:
```diff
@@ -786,9 +783,138 @@ for [future :unicorn:][future multiple tables] use and must be 0 in the MVP.
+
+## Atomic compare exchange operators ([described here](Semantics.md#atomic-memory-accesses))
+
+| `i32.atomic_cmpxchg` | `0xe61c` | `memory immediate` | atomic compare exchange with memory |
```
The formatting on this table is broken, possible because of the lack of a heading.

binji replied:
Oops, thanks!

sunfishcode commented on BinaryEncoding.md:
```diff
@@ -786,9 +783,138 @@ for [future :unicorn:][future multiple tables] use and must be 0 in the MVP.
+| `i32.atomic_xor8_s` | `0xe612` | `4` `memory immediate` | atomic xor with memory |
+| `i32.atomic_xor8_u` | `0xe613` | `4` `memory immediate` | atomic xor with memory |
+| `i32.atomic_xor16_s` | `0xe614` | `4` `memory immediate` | atomic xor with memory |
+| `i32.atomic_xor16_u` | `0xe615` | `4` `memory immediate` | atomic xor with memory |
```
Names like `i32.atomic_xor16_s` are a little awkward with `_s`/`_u` apparently associated with otherwise sign-agnostic operators. What would you think of any of the following conventions?

 - `i32.atomic16_s.xor`
 - `i32.atomic_rmw16_s.xor`
 - `i32.atomic.rmw16_s.xor`   (in this case, following atomic with `.` could be a broader convention)

These would make it clear that the xor isn't signed itself; it's just the loaded value from the rmw that's being sign-extended. This isn't super important with the current opcodes, but could be in the future if wasm ever wants atomic rotate, popcnt, or other operators which care about the number of bits in the actual arithmetic part.

Such a clarification would also align with an idea anticipated [here](https://github.com/WebAssembly/design/blob/master/FutureFeatures.md#integer-overflow-detection) concerning the possibility of adding a real `add_s` and similar operators (which could trap or do other things on overflow) in the future.

binji replied:
I like `i32.atomic.rmw16_s.xor`, it nicely matches the proposed opcode layout too. Though I wonder about `xchg` and `cmpxchg`:
`i32.atomic.rmw16_s.xchg`? And as @jfbastien mentioned, cmpxchg isn't really RMW, so `i32.atomic.cmpxchg16_s`? And I suppose for the non-extending forms:
`i32.atomic.rmw.xor`?

sunfishcode replied:
`i32.atomic.cmpxchg16_s` and `i32.atomic.rmw.xor` seem to fit the pattern nicely.

binji replied:
Ok, done

lukewagner commented:
@flagxor I'm not sure what would be better without knowing what the intended process is for proposing and making changes.  Since threads are a high priority, we could just cheat and carry on like we did pre-MVP.

lukewagner commented on BinaryEncoding.md:
```diff
@@ -786,9 +783,140 @@ for [future :unicorn:][future multiple tables] use and must be 0 in the MVP.
+| Name | Opcode | Immediate | Description |
+| ---- | ---- | ---- | ---- |
+| `i32.atomic_load` | `0xe628` | `memory immediate` | atomic load from memory |
+| `i64.atomic_load` | `0xe629` | `memory immediate` | atomic load from memory |
```
For all these load/store ops, could we reuse the existing load/store ops, adding a "consistency" annotation to the `memory_immediate.flags`?

binji replied:
Yes, and I believe that's what was discussed before. But I had a thought it might be nice to have all atomic/threading operators use the same prefix, which led me down this route. One additional nicety is that the opcodes for atomic load/store operators match the non-atomic ones, e.g. `i32.atomic_load` is `0xe628` and `i32.load` is `0x28`, etc.

Since the opcode specifies that it is atomic, this also means that we don't need a special "non-atomic" value for the memory immediate for RMW ops too.

But I'm not married to the idea, just thought it was kinda neat. :-)

lukewagner replied:
Hm, those are fair points.  I guess we are already adding a bunch of operators, so it's not a huge difference, and there would be a natural extension for using the atomic prefix on global ops (when we eventually allow shared globals).

jfbastien replied:
I'm partial to calling the "atomic" prefix the "lock" prefix, and numbering it `0xF0`.

binji replied:
@jfbastien can't tell if serious...

sunfishcode commented on Semantics.md:
```diff
@@ -157,6 +162,146 @@ Store operators do not produce a value.
+
+| Name | Read | Modify | Write |
+| ---- | ---- | ---- | ---- |
+| `i32.atomic_cmpxchg8_s` | 1 byte, sign-extended i8 to i32 | nop | (conditional) 1 byte, wrapped from i32 to i8 |
```
One detail that's not clear from this presentation is whether the comparison is the sign-extended value with the full 32 bits of the expected value, or the expected value wrapped to 8 bits with the 8-bit loaded value.

binji replied:
I tried to clarify, ptal

AndrewScheidecker commented:
What about atomic operations on globals?

lukewagner commented:
@AndrewScheidecker I think first we need to define a new `WebAssembly.Global` object for mutable globals that can be imported, exported and marked `shared` (symmetric to `Memory`); but after that it makes sense to have atomic global operations.

sunfishcode commented on Semantics.md:
```diff
@@ -157,6 +162,146 @@ Store operators do not produce a value.
+
+| Name | Read | Modify | Write |
+| ---- | ---- | ---- | ---- |
+| `i32.atomic.rmw8_s.add` | 1 byte, sign-extended i8 to i32 | sign-agnostic addition | 1 byte, wrapped from i32 to i8 |
```
The cmpxchg clarification looks good. So good, that I'm now wondering about a Return column for the rmw table too :-). This is admittedly a really minor nit at this point because none of the current rmw operators care whether their inputs are extended or wrapped, but if we add other rmw operators in the future, they could potentially care.

rossberg-chromium commented on Semantics.md:
```diff
@@ -678,6 +823,50 @@ outside the range which rounds to an integer in range) traps.
+## Thread operators
+
+  * `is_lock_free`: If the atomic step of an atomic primitive (see
+    [Atomic Memory Accesses](#atomic-memory-accesses)) on a datum of size N
```
Where is N coming from, is that an i32 operand?

binji replied:
Done.

rossberg-chromium commented on Semantics.md:
```diff
@@ -157,6 +162,146 @@ Store operators do not produce a value.
+  * `f32.atomic.store`: (no conversion) atomically store 4 bytes
+  * `f64.atomic.store`: (no conversion) atomically store 8 bytes
+
+Atomic read-modify-write (RMW) operators atomically read a value from a memory
```
Can you specify in which order the operands are expected on the stack for these instructions?

binji replied:
I couldn't find another place in the doc to crib from, so I just specified it by describing the . Should I be more explicit?

rossberg-chromium commented on Semantics.md:
```diff
@@ -157,6 +162,146 @@ Store operators do not produce a value.
+| `i64.atomic.rmw32_u.xchg` | 4 bytes | nop | 4 bytes | zero-extended i32 to i64 |
+| `i64.atomic.rmw.xchg` | 8 bytes | nop | 8 bytes | as i64 |
+
+The atomic compare exchange operators take three operands: a linear memory
```
Nit: the linear memory is an immediate, not an operand.

binji replied:
Whoops, I meant address. :-)

rossberg-chromium commented on Semantics.md:
```diff
@@ -678,6 +823,50 @@ outside the range which rounds to an integer in range) traps.
+change. It is a validation error to use these operators on non-shared linear
+memory.
+
+The wait operators take three operands: a linear memory index, an expected
```
Nit: as above, I suppose the memory is an immediate.

rossberg-chromium commented on BinaryEncoding.md:
```diff
@@ -786,9 +783,140 @@ for [future :unicorn:][future multiple tables] use and must be 0 in the MVP.
+
+| Name | Opcode | Immediate | Description |
+| ---- | ---- | ---- | ---- |
+| `is_lock_free` | `0xe600` | | |
```
Can we allocate prefix opcodes at the end of the opcode space instead of the middle? So `0xff` for threads being the first such proposal.

binji replied:
Done.

rossberg-chromium commented on Semantics.md:
```diff
@@ -678,6 +823,50 @@ outside the range which rounds to an integer in range) traps.
+
+### Wake
+
+The wake operators take two operands: a linear memory index, and a wake count.
```
Nit: same here.

rossberg-chromium commented on BinaryEncoding.md:
```diff
@@ -786,9 +783,140 @@ for [future :unicorn:][future multiple tables] use and must be 0 in the MVP.
+| `i32.wait` | `0xe601` | `offset` | |
+| `i64.wait` | `0xe602` | `offset` | |
+| `i32.wake` | `0xe603` | `offset` | |
+| `i64.wake` | `0xe604` | `offest` | |
```
Don't these all have memory immediates?

binji replied:
Done.

rossberg-chromium commented on BinaryEncoding.md:
```diff
@@ -786,9 +783,140 @@ for [future :unicorn:][future multiple tables] use and must be 0 in the MVP.
+| `i64.atomic.rmw16_u.xor` | `0xe619` | `4` `memory immediate` | atomic xor with memory |
+| `i64.atomic.rmw32_s.xor` | `0xe61a` | `4` `memory immediate` | atomic xor with memory |
+| `i64.atomic.rmw32_u.xor` | `0xe61b` | `4` `memory immediate` | atomic xor with memory |
+| `i32.atomic.rmw.xchg` | `0xe610` | `5` `memory immediate` | atomic exchange with memory |
```
Is xchg actually a RMW operation? Seems like it could be its own group, with names like `i32.atomic.xchg16_s`, in line with `cmpxchg`.

rossberg-chromium commented on Semantics.md:
```diff
@@ -157,6 +162,146 @@ Store operators do not produce a value.
+| `i64.atomic.rmw32_s.xor` | 4 bytes | 32-bit sign-agnostic bitwise exclusive or | 4 bytes | sign-extended i32 to i64 |
+| `i64.atomic.rmw32_u.xor` | 4 bytes | 32-bit sign-agnostic bitwise exclusive or | 4 bytes | zero-extended i32 to i64 |
+| `i64.atomic.rmw.xor` | 8 bytes | 64-bit sign-agnostic bitwise exclusive or | 8 bytes | as i64 |
+| `i32.atomic.rmw8_s.xchg` | 1 byte | nop | 1 byte | sign-extended i8 to i32 |
```
Is xchg actually a RMW operation? It seems like it would fit better with the `cmpxchg` group -- it's an unconditional version of it. With analogous names like `i32.atomic.xchg16_s`. Or as its own group.

jfbastien replied:
xchg is usually considered an RMW operation.

binji replied:
I'm fine with it either way, but it probably shouldn't be grouped with `cmpxchg` since those have 3 operands.

rossberg-chromium replied:
I guess I'm wondering by what criterion xchg is RMW but cmpxchg isn't.

Bikeshedding territory, but why the complicated names in the first place? Wouldn't i32.atomic.add8_s etc work just fine? Sure it's RMW, but that's kind of implied.

sunfishcode replied:
`cmpxchg` (compare-and-swap) is fundamentally [more powerful](http://www.cs.yale.edu/homes/aspnes/pinewiki/WaitFreeHierarchy.html) than other RMW operations. Between that, and having a distinct type signature, it commonly gets its own category, even if it is also an RMW operation.

`cmpxchg` isn't just a conditional version of `xchg`. For example, an abstract formal semantics wouldn't be able to reduce `cmpxchg` to `xchg` when the compare succeeds; that would imply the comparison is a separate step from the swap, breaking atomicity.

`add8_s` under wasm's existing naming scheme would seem to imply that the `add`, `or`, `and`, `xor`, etc., are signed/unsigned, which is inconsistent with the existing operators.

rossberg-chromium replied:
@sunfishcode, interesting, thanks!

Regarding the naming, I see your argument. My only remaining quibble is that, even being in different expressive categories, it would still be nice to have a somewhat more symmetric naming scheme between cmpxchg and the other RMW operators.

sunfishcode replied:
`i32.atomic.rmw8_s.cmpxchg` would be the maximally symmetric convention, it seems. I don't think it's critical, but I wouldn't be opposed to it.

binji replied:
Done.

sbc100 commented on Semantics.md:
```diff
@@ -157,6 +162,146 @@ Store operators do not produce a value.
+The atomic compare exchange operators take three operands: a linear memory
+index, an `expected` value, and a `replacement` value. The operators
+conditionally store the `replacement` value, but only if the `loaded` value is
+equal to the `expected` value.
```
The logic of this sentence seems a little strange to me.  "X conditionally does Y, but only when Z"..  I think what you are trying to say is that its always conditional, and then describe said condition.

binji replied:
Done.

jfbastien commented:
I think we've got a great opportunity to call the "atomic" prefix the "lock" prefix, and number it `0xF0`.

syg commented:
> Yeah, I was considering just referencing the ecma262 memory model, but it's not immediately obvious to a reader how it would map.

@binji Happy to help if you bring me up to speed about spec conventions used here.

jfbastien commented on BinaryEncoding.md:
```diff
@@ -786,6 +950,18 @@ for [future :unicorn:][future multiple tables] use and must be 0 in the MVP.
+| `i64.wake` | `0xc3` | `offest` | SAB like wake, takes an index + `offset` |
+| `i32.thread_create` | `0xc4` | `table` | Creates a thread, stores thread in `table` at an i32 index. Returns 0 on success or non-zero otherwise. |
+| `i32.thread_join` | `0xc5` | `table` | Waits to join to thread stored at an i32 index in `table`. Returns 0 on success or non-zero otherwise. |
+| `thread_exit` | `0xc6` | | Halts current thread. |
```
It would be good to document this! Is the intent to start off with workers, mimic JS' API, and then add better capabilities later?

binji commented:
@jfbastien isn't that a bit biased? ;-)

@syg not 100% on conventions here, since it's informal. For memory model stuff, I'd guess it would be best to write a new markdown file that we could reference here.

sunfishcode commented:
Can we specify a set of sizes for which `is_lock_free` is guaranteed to return true? This was [discussed earlier](https://github.com/WebAssembly/design/issues/300), and it seems that 1, 2, and 4 should be lock free on all relevant platforms. Such a guarantee allows applications to avoid having fallback paths in common situations. And, these would support eg. C++'s `std::atomic::is_always_lock_free`.

Another topic specific to wasm is alignment for atomics. It seems that at least, `is_lock_free` doesn't describe misaligned atomic accesses, though one could contemplate stronger constraints as well, such as having misaligned atomics trap.

Is this the best place for these topics, or should these be addressed in the (upcoming) memory model document?

jfbastien commented:
@sunfishcode 
> Can we specify a set of sizes for which `is_lock_free` is guaranteed to return true? This was [discussed earlier](https://github.com/WebAssembly/design/issues/300), and it seems that 1, 2, and 4 should be lock free on all relevant platforms. Such a guarantee allows applications to avoid having fallback paths in common situations. And, these would support eg. C++'s `std::atomic::is_always_lock_free`.

I'd like to match exactly what JavaScript guarantees. A mismatch would be akin to raw C++ programs talking through shared memory with different compiler flags (e.g. `is_always_lock_free` differing!). Fraught with peril because code will statically decide to behave differently (say, use a lock-free datastructure versus a plain old lock).

> Another topic specific to wasm is alignment for atomics. It seems that at least, `is_lock_free` doesn't describe misaligned atomic accesses, though one could contemplate stronger constraints as well, such as having misaligned atomics trap.

And SAB doesn't specify alignment of atomics because the array buffer accesses are aligned, right? We're in novel territory here, and we know there's significant hardware difference in what outcome can be observed.

It would be good to lay out different options and perform some Science. At the end of the day alignment of atomics is a non-feature which we *must* spec, but I don't really care *how* we spec it. Determinism is best, unless the cost is too high.

For now I'd lean towards trap as @sunfishcode suggests, but I can be moved away.

> Is this the best place for these topics, or should these be addressed in the (upcoming) memory model document?

Does the memory model address wasm-specific issues?

Agreed though, maybe we want this entire change to live in its own document for now, and merge its sub-parts once we've reached consensus.

binji commented:
> I'd like to match exactly what JavaScript guarantees.

JavaScript only guarantees that 4-byte is lock free.

> And SAB doesn't specify alignment of atomics because the array buffer accesses are aligned, right?

Right, and you can't use atomics on a `DataView`.

> For now I'd lean towards trap as @sunfishcode suggests, but I can be moved away.

+1

> Agreed though, maybe we want this entire change to live in its own document for now, and merge its sub-parts once we've reached consensus.

Not a bad idea, something like a "Threads-Proposal.md" that specifies how the rest of the files would change? Is that nicer than having a "thread-proposal" branch that we review PRs on?

jfbastien commented:
>> Agreed though, maybe we want this entire change to live in its own document for now, and merge its sub-parts once we've reached consensus.
>
> Not a bad idea, something like a "Threads-Proposal.md" that specifies how the rest of the files would change? Is that nicer than having a "thread-proposal" branch that we review PRs on?

Maybe something that won't change names once stabilized, so URLs on the internets don't break.

sunfishcode commented:
> I'd like to match exactly what JavaScript guarantees. A mismatch would be akin to raw C++ programs talking through shared memory with different compiler flags (e.g. is_always_lock_free differing!). Fraught with peril because code will statically decide to behave differently (say, use a lock-free datastructure versus a plain old lock).

If, within a given engine, JS `Atomics.isLockFree` agrees with WebAssembly `is_lock_free` for all values, then there is no problem. On the other hand, if, within a given engine, JS `Atomics.isLockFree` ever disagrees with WebAssembly `is_lock_free`, for any reason, it'd cause all the problems, even if without static guarantees, because JS code and WebAssembly code would *dynamically* pick different datastructures.

So adding a static guarantee doesn't seem to introduce any new peril.

lars-t-hansen commented:
@sunfishcode 

> Can we specify a set of sizes for which is_lock_free is guaranteed to return true? This was discussed earlier, and it seems that 1, 2, and 4 should be lock free on all relevant platforms. Such a guarantee allows applications to avoid having fallback paths in common situations. And, these would support eg. C++'s std::atomic::is_always_lock_free.

I think the matter has been resolved but for the record:

It is arguable that 1 and 2 are not lock free on all relevant platforms.  MIPS has only 4-byte LL/SC so if you want a 1-byte lock-free atomic you must use LL/SC on the word that contains that byte.  Now the other bytes of that word may be used for other, unrelated variables that are not necessarily atomic, and we don't control that.  In this situation we're only good if a concurrent non-atomic update to any of those other bytes will cause the SC to fail and the LL/SC loop to retry.  It's not obvious from the architecture manuals that I have seen that this is so (though ISTR that I found some hints in that direction).

> Another topic specific to wasm is alignment for atomics. It seems that at least, is_lock_free doesn't describe misaligned atomic accesses, though one could contemplate stronger constraints as well, such as having misaligned atomics trap.

Copy atomicity is also compromised by misaligned atomic accesses (but so is copy atomicity on regular accesses).  I don't know how much that matters.  The JS memory model goes out of its way to specify copy atomicity in some situations, but assumes alignment.

IMO unaligned atomics should just trap, until we know for sure that we can reasonably support atomics that don't.

sunfishcode commented:
@lars-t-hansen Thanks for the explanation. Clang is currently written in a way that does not permit it to consider this, and it therefore doesn't consider it on MIPS or any other of its targets. I have now filed https://reviews.llvm.org/D31516 with a patch to Clang which introduces logic in clang to allow it to consider this. Hopefully either the patch will be accepted, or we will learn more.

binji commented:
Thanks, @sunfishcode for looking into this. In the meantime, I've updated this PR to specify that 1- and 2-byte atomics are not guaranteed lock-free. I've also added some text about misaligned atomics.

jfbastien commented on Semantics.md:
```diff
@@ -678,6 +829,51 @@ outside the range which rounds to an integer in range) traps.
+woken.
+
+  * `i32.wake`: wake up to N threads waiting on the given address via `i32.wait`
+  * `i64.wake`: wake up to N threads waiting on the given address via `i64.wait`
```
I think we only want the 32-bit variants of wait / wake.

jfbastien replied:
How do I "wake all"? In JS I can pass infinity. It seems silly to pass UINT_MAX.

binji replied:
Done.

binji replied:
Yeah, this is a little clunky. I made it a signed `i32`, where < 0 means wake all. Using an `f64` for the wake count seemed really odd. Perhaps it could be a separate opcode?

jfbastien replied:
Yes, separate `wakeAll` sounds good.

jfbastien commented on Semantics.md:
```diff
@@ -678,6 +829,51 @@ outside the range which rounds to an integer in range) traps.
+    primitive (see [Atomic Memory Accesses](#atomic-memory-accesses)) on a
+    datum of size `N` bytes will be performed without the agent acquiring a
+    lock outside the `N` bytes comprising the datum, then return `1`.
+    Otherwise, return `0`.
```
Is `N` an immediate? It's not clear.

What type?

Are there invalid values for `N`?

Can we express the `constexpr`-ness of this? SAB had some wording somewhere about not changing between visible agents or something.

Are `wait` and `wake` atomic primitives? Because this wording would affect them and they definitely aren't lock-free. Currently the headings seem to say "no", it should remain that way.

binji replied:
AIUI, we use `operand` to mean stack value, correct? I could be more explicit if desired, though. Tried to clear up the other stuff, PTAL.

jfbastien commented on Semantics.md:
```diff
@@ -678,6 +829,51 @@ outside the range which rounds to an integer in range) traps.
+The wait operators take three operands: an address operand, an expected
+value, and a relative timeout in milliseconds as an f64. The timeout must be
+non-negative, and can also be positive infinity. If the timeout is positive
+infinity then the timeout will never expire.
```
Why disallow negative values? A timer would just be expired if it were negative. Otherwise, specify the effects of a negative timer.

What of NaN?

binji replied:
Good points, I made this value match the behavior of JS's `Atomics.wait`.

jfbastien commented on Semantics.md:
```diff
@@ -678,6 +829,51 @@ outside the range which rounds to an integer in range) traps.
+The wait operation begins by performing an atomic load from the given address.
+If the loaded value is not equal to the expected value, the operator returns 1
+("not-equal"). If the values are equal, the thread is suspended. If the thread
+is woken (by execution of the corresponding "wake" operator on another thread),
```
What if the wake overlaps but isn't aligned?

binji replied:
Done.

jfbastien commented on Semantics.md:
```diff
@@ -678,6 +829,51 @@ outside the range which rounds to an integer in range) traps.
+non-negative, and can also be positive infinity. If the timeout is positive
+infinity then the timeout will never expire.
+
+The wait operation begins by performing an atomic load from the given address.
```
Implication here is it can fault if out-of-bounds, and can't be misaligned?

jfbastien replied:
What's the type of the load, and `expected`?

binji replied:
Done.

binji replied:
Done.

jfbastien commented on Semantics.md:
```diff
@@ -678,6 +829,51 @@ outside the range which rounds to an integer in range) traps.
+### Wake
+
+The wake operators take two operands: an address operand, and a wake count. The
+operation will wake a maximum of "wake count" threads that are waiting on the
```
The wording can be interpreted as saying: there can be 1000 waiters, it is valid to wake none of them, ever.

binji replied:
Good point, fixed.

jfbastien commented on Semantics.md:
```diff
@@ -678,6 +829,51 @@ outside the range which rounds to an integer in range) traps.
+infinity then the timeout will never expire.
+
+The wait operation begins by performing an atomic load from the given address.
+If the loaded value is not equal to the expected value, the operator returns 1
```
Return is `i32`?

binji replied:
Done.

jfbastien commented on Semantics.md:
```diff
@@ -678,6 +829,51 @@ outside the range which rounds to an integer in range) traps.
+
+### Wake
+
+The wake operators take two operands: an address operand, and a wake count. The
```
Wake count is `i32`, interpreted as unsigned?

jfbastien commented on BinaryEncoding.md:
```diff
@@ -786,9 +783,140 @@ for [future :unicorn:][future multiple tables] use and must be 0 in the MVP.
+| `i32.wait` | `0xff01` | `memory immediate` | |
+| `i64.wait` | `0xff02` | `memory immediate` | |
+| `i32.wake` | `0xff03` | `memory immediate` | |
+| `i64.wake` | `0xff04` | `memory immediate` | |
```
I think we only want the 32-bit variants of wait / wake.

jfbastien replied:
What's the memory immediate?

jfbastien replied:
Do we intend wait / wake to have sequentially-consistent semantics, or acquire (wait) / release (wake) sematics?

binji replied:
Done.

binji replied:
The JS behavior is sequentially-consistent, so we should match, I think.

jfbastien commented on BinaryEncoding.md:
```diff
@@ -786,9 +783,140 @@ for [future :unicorn:][future multiple tables] use and must be 0 in the MVP.
+
+| Name | Opcode | Immediates | Description |
+| ---- | ---- | ---- | ---- |
+| `i32.atomic.rmw.add` | `0xff10` | `0` `memory immediate` | atomic add to memory |
```
Why use `memory immediate` to denote the operation (add / sub / etc)? Why not use another opcode instead? It would be helpful to do a map of opcode extension space usage here, similar to what was done for the MVP opcodes. Is the intent that `0xff` is the "atomics / threads" prefix? I still think we should use `0xf0`, it's free!

binji replied:
> Why use memory immediate to denote the operation (add / sub / etc)? Why not use another opcode instead?

I considered using separate opcodes, but it eats up the opcode space pretty quickly since we have 12 variants per operation. I also kind of like that all the rmw ops are grouped together, but I suppose that's just personal preference.

> It would be helpful to do a map of opcode extension space usage here, similar to what was done for the MVP opcodes.

Sorry, do you mean mapping the usage of the atomics prefix space? Or mapping the usage of the remaining opcode space?

> I still think we should use 0xf0, it's free!

I don't have a preference for "atomics/threads" prefix, `0xf0` is just as good as anything else as far as I'm concerned. But it does seem like adding opcode prefixes starting at the end and going backward is logical. I suppose we could argue for 0xf0 by saying all prefixes start with `0xf_` and this is the first one.

jfbastien replied:
>> It would be helpful to do a map of opcode extension space usage here, similar to what was done for the MVP opcodes.
>
> Sorry, do you mean mapping the usage of the atomics prefix space? Or mapping the usage of the remaining opcode space?

The former.

> I don't have a preference for "atomics/threads" prefix, 0xf0 is just as good as anything else as far as I'm concerned. But it does seem like adding opcode prefixes starting at the end and going backward is logical. I suppose we could argue for 0xf0 by saying all prefixes start with 0xf_ and this is the first one.

🎉

jfbastien commented on Semantics.md:
```diff
@@ -112,8 +112,13 @@ or [defined inside the module](Modules.md#linear-memory-section). After import
+threads of execution. The shared memory can be imported or defined in the
+module. It is a validation error to attempt to import shared linear
+memory if the module's memory import doesn't specify that it allows shared
+memory.
```
There are two places where we can mark a memory as shared:

1. In the `WebAssembly.Memory` constructor. You need to update JS.md for this.
2. In [`memory_type`](https://github.com/WebAssembly/design/blob/master/BinaryEncoding.md#memory_type) (for both import and memory sections). This looks like we don't have room for expansion though?

You also need to update JS.md to specify that `.buffer` will return a `SharedArrayBuffer`.

We also must specify what happens when you `.grow` or use `grow_memory`.

jfbastien commented:
How does one share a `WebAssembly.Memory`, which is marked as shared, across workers? Structured clone?

binji commented:
Yeah, I was hoping to get away with adding the easy stuff before diving in to all the other thread stuff. But it doesn't really make sense to add these primitives until we have a good base to build on top of. I'll work on a `Threads.md` next for this.

AndrewScheidecker commented:
I was trying to figure out if it's possible to implement the `wait`/`wake` ops using the Windows API's `WaitOnAddress`/`WakeByAddressSingle`/`WakeByAddressAll`, and found some interesting caveats in how they are currently implemented: [search on this page for WaitOnAddress](http://www.cbloom.com/rambles.html), which references [this post on the subject](https://blogs.msdn.microsoft.com/oldnewthing/20160826-00/?p=94185).

However, it seems like if `wait`/`wake` are specified loosely enough to be implemented using the system-level APIs in Windows, then the only robust way to use `wake` will be to wake all waiting threads. That's still useful, but maybe there should just be a `wake_all` op for now.

jfbastien commented:
> I was trying to figure out if it's possible to implement the `wait`/`wake` ops using the Windows API's `WaitOnAddress`/`WakeByAddressSingle`/`WakeByAddressAll`, and found some interesting caveats in how they are currently implemented: [search on this page for WaitOnAddress](http://www.cbloom.com/rambles.html), which references [this post on the subject](https://blogs.msdn.microsoft.com/oldnewthing/20160826-00/?p=94185).
>
> However, it seems like if `wait`/`wake` are specified loosely enough to be implemented using the system-level APIs in Windows, then the only robust way to use `wake` will be to wake all waiting threads. That's still useful, but maybe there should just be a `wake_all` op for now.

We should match the SharedArrayBuffer spec at a minimum, because JS and wasm can wait / wake each other, and write to the same atomic locations. I'll punt implementability concerns to SAB: if MS folks were OK with SAB they'll be OK with the same API here.

@binji This embedder/embedded wake/wait should be pointed out in the wording.

jfbastien commented on BinaryEncoding.md:
```diff
@@ -789,9 +789,14 @@ for [future :unicorn:][future multiple tables] use and must be 0 in the MVP.
+| Name | Type | Description |
+| ---- | ---- | ---- |
+| flags | `varuint32` | a bitfield which currently must be 0 |
+| offset | `varuint32` | the value of the offset |
```
Is it useful to have offsets on wake / wait? I'm not sure, but open to being convinced for / against. Seems to match with other memory ops?

binji replied:
I figure it makes sense for consistency. And is convenient for the same reason that it is for other memory accesses, so you can have a pointer to a struct and use that for wait/wake instead of having to add it in yourself.

jfbastien replied:
Yeah sounds fine. Just figured I'd ask :)

jfbastien commented on Semantics.md:
```diff
@@ -831,29 +831,39 @@ outside the range which rounds to an integer in range) traps.
+    acquiring a lock outside the `N` bytes comprising the datum, then return
+    `1`. Otherwise, return `0`. Once the value of `is_lock_free` for a given
+    value `N` has been observed for any [agent](agent) in an
+    [agent cluster](agent cluster) , it cannot change.
```
Extra space here.

jfbastien commented on Semantics.md:
```diff
@@ -831,29 +831,39 @@ outside the range which rounds to an integer in range) traps.
-thread wakes this one, this operator returns 2 ("timed-out").
+| `timeout` value | Behavior |
+| ---- | ---- |
+| `timeout` <= 0 | Expires immediately |
```
"immediately": is it guaranteed synchronous, or can it wait until the next runloop?

binji replied:
Good question, it's not clear to me from the ES spec what the behavior is. Seems like synchronous is useful, but it's also often good to have synchronous vs. asynchronous behavior be consistent for all values of an operator. What do you think?

jfbastien replied:
I defer to @lars-t-hansen / @syg on this.

jfbastien commented on Semantics.md:
```diff
@@ -862,18 +872,24 @@ thread wakes this one, this operator returns 2 ("timed-out").
 
-  * `i32.wake`: wake up to N threads waiting on the given address via `i32.wait`
-  * `i64.wake`: wake up to N threads waiting on the given address via `i64.wait`
+  * `i32.wake`: wake up `wake count` threads waiting on the given address via `i32.wait`
```
I think you went a bit too far: if there aren't `wake count` waiters then you just wake all of them. The SAB algorithm does this IIUC. We could use `min(wake count , waiters)` or something?

binji replied:
Oops, I updated the text above but this one needs to be fixed too.

jfbastien commented:
Actually I talked with @pizlonator and it may be useful to keep i32 + i64 variants of wake / wait: if we specify that only the address, not the size, participate in realizing wake / wait then it's super convenient to have both. I was wary of how we'd specify overlap, but based on address just works.

jfbastien commented on Semantics.md:
```diff
@@ -848,6 +848,12 @@ memory. They are non-atomic and have sequentially consistent memory ordering.
 
+The embedder is also permitted to suspend or wake a thread. In those cases,
+these operators will behave as though the equivalent wake or wait operation was
+performed by the execution engine.
```
This goes both ways: a wake will also wake waiters from the embedder.

binji replied:
Right, I guess that wasn't clear from the wording here. Let me try again :)

AndrewScheidecker commented:
@jfbastien I missed [the SAB wait/wake](https://tc39.github.io/ecma262/#sec-atomics.wait) somehow when I looked before.

Incidentally, I believe I found an erroneous assertion in the SAB spec. In [`Atomics.wait`](https://tc39.github.io/ecma262/#sec-atomics.wait), step 18 performs `RemoveWaiter` after `Suspend` completes. `RemoveWaiter` asserts that the agent being removed is in the `WaiterList`. However, if `Suspend` returned because a call to `Atomics.wake` called `WakeWaiter`, the agent will have been removed from the `WaiterList` already by step 11 of [`Atomics.wake`](https://tc39.github.io/ecma262/#sec-atomics.wake).

The SAB wait/wake seems to be specified in a way that can't use any of the existing OS-level APIs. e.g. it specifies wake order where Linux futexes do not, and it specifies that a `wait` call with infinite timeout will *only* be woken by a `wake` call that occurs after the `wait` call, which isn't necessarily the case on Windows.

My original thought was that it would be useful to spec the futex stuff in a way that can use the OS-level APIs to synchronize with native processes through shared memory, but I see now that's futile. However, it does suggest that some higher-level synchronization primitives should be exposed that *can* plug into the OS-level primitives. That is probably beyond the scope of the initial threading extensions.

binji commented:
@AndrewScheidecker I think the spec is correct; it looks like `RemoveWaiters` creates a new list with one of the waiters removed, but it doesn't modify the original list.

AndrewScheidecker commented:
@binji The RemoveWaiters spec says:
> 1. Assert: The calling agent is in the critical section for WL.
> 2. Let L be a new empty List.
> 3. Let S be a reference to the list of waiters in WL.
> 4. Repeat, while c > 0 and S is not an empty List,
>     a. Let W be the first waiter in S.
>     b. Add W to the end of L.
>     c. Remove W from S.
>     d. Subtract 1 from c.
> 5. Return L.

I would interpret step 3's "Let S be a reference to the list of waiters in WL" to mean that when step 4c removes W from S, it's removing W from WL's list of waiters.

binji commented:
@AndrewScheidecker Yes, I see what you mean. Filed a bug here: https://github.com/tc39/ecma262/issues/861

binji commented:
Moved everything to a single Threads.md file, curious what everyone thinks.

jfbastien commented on Threads.md:
```diff
@@ -0,0 +1,403 @@
+
+The wake and wait operators are optimizations over busy-waiting for a value to
+change. It is a validation error to use these operators on non-shared linear
+memory. They are non-atomic and have sequentially consistent memory ordering.
```
They're not in the "atomic" section of operators, but the "read' operation from wait is atomic.

jfbastien commented on Threads.md:
```diff
@@ -0,0 +1,403 @@
+
+| Name | Opcode | Immediates | Description |
+| ---- | ---- | ---- | ---- |
+| `i32.atomic.rmw.add` | `0xff10` | `0` `memory immediate` | atomic add to memory |
```
I still find this kinda weird: there's the memory immediate as above, but here there's an opcode number as well. If we carve off all of `0xf0` for atomics then we've got plenty of space, and we'd use a byte less.

binji replied:
Maybe you're right, I made a gist mapping out the opcode space, and we still have quite a bit of room if we expand these out: https://gist.github.com/binji/0d498bbe16173e7fad18cfb39897e2b0

jfbastien replied:
And then every op we encode saves a byte, right?

binji replied:
Yes, although I'd guess that atomic RMW ops are relatively rare so that extra byte is probably not a big deal.

jfbastien replied:
Yeah I agree that's not a big criteria. It just seems weird to have a modifier since we don't do this elsewhere. At that point all atomics could be a single op with a modifier, instead of using prefixes.

binji replied:
OK, done.

lars-t-hansen commented on Threads.md:
```diff
@@ -0,0 +1,403 @@
+| `timeout` <= 0 | Expires immediately |
+| 0 < `timeout` < Positive infinity | Expires after `timeout` milliseconds |
+| Positive infinity | Never expires |
+| NaN | Never expires |
```
Given how heavyweight wait is, my slight preference would probably be for negative and NaN timeouts to trap, but I do see the significant appeal of making timeouts in the past expire immediately.

NaN though - surely NaN here would indicate an ill-computed timeout and a trap would be as meaningful as would a range error on a heap access?

jfbastien replied:
NaN is free to handle if you do: current < timeout. It'll be false if timeout is NaN.

lars-t-hansen replied:
The cost of the operation seems like a non-argument for `wait` more than for almost any other functionlity I can imagine wasm having.  I would also expect implementations to convert an FP argument to some integer type and compute with that, so the NaN has to be handled specially in any case.

binji replied:
I actually just copied the behavior from `Atomics.wait` for NaN. :-)

lars-t-hansen replied:
Hrmpf.  Well, fair enough.  In my (feeble) defense the NaN behavior there was chosen to mimic existing APIs, and IMO we should do better, but I'll stop arguing :)

rossberg-chromium commented:
@jfbastien, can you motivate why you insist on 0xf0? It seems strictly worse than starting from the end.

jfbastien commented:
> @jfbastien, can you motivate why you insist on 0xf0? It seems strictly worse than starting from the end.

Because it's fun to match x86, and starting from `0xf0` and counting up is pretty neat for extensions.

rossberg-chromium commented:
@jfbastien, I don't see the relevance of matching x86 -- if anything I'd avoid that for the sake of avoiding bias.

jfbastien commented:
> @jfbastien, I don't see the relevance of matching x86 -- if anything I'd avoid that for the sake of avoiding bias.

When picking arbitrary numbers, fun becomes the bikeshed differentiator. 🐴

rossberg-chromium commented:
@jfbastien, minimizing fragmentation in the opcode space > fun ;)

jfbastien commented:
> @jfbastien, minimizing fragmentation in the opcode space > fun ;)

That's already something that was given up on quite a while ago. Witness `0x06`–`0x0a` and other gaps.

This is a super simple, inconsequential, silly thing. There's goodness in designs having elegance, and there's also goodness in designs having some witticism. People in a few years will look at this silly thing, say "cute, designers knew how to have fun" and smile.

lukewagner commented on Threads.md:
```diff
@@ -0,0 +1,403 @@
+threads of execution. The shared memory can be imported or defined in the
+module. It is a validation error to attempt to import shared linear
+memory if the module's memory import doesn't specify that it allows shared
+memory.
```
It'd be good to flesh this out a bit more, including [these sorts of details](https://github.com/WebAssembly/design/issues/950#issuecomment-272360540).

binji replied:
I've added some of the details, PTAL (still no JS.md stuff yet)

jfbastien commented:
We need to specify:

* What happens with `grow_memory` and `WebAssembly.Memory.prototype.grow`.
* What happens to `instance.export.memory.buffer` that are in JS.
* Whether `current_memory` is sequentially-consistent.

binji commented on Threads.md:
```diff
@@ -0,0 +1,403 @@
+# Threads
+
+This page describes a proposal for the post-MVP
+[threads feature][future threads].
```
how about :atom_symbol: (`:atom_symbol:`)

binji commented on Threads.md:
```diff
@@ -0,0 +1,403 @@
+  * `f64.atomic.store`: (no conversion) atomically store 8 bytes
+
+Atomic read-modify-write (RMW) operators atomically read a value from an
+address , modify the value, and store the resulting value to the same address.
```
Done.

binji commented:
OK, I've added a little bit here about agents/agent clusters, mostly copied from https://tc39.github.io/ecma262/#sec-agents. Not sure how much detail we want to go into here. Gonna write up some JS.md stuff next.

jfbastien commented on Threads.md:
```diff
@@ -31,15 +57,32 @@ Alignment of non-atomic accesses affects performance as follows:
+...
+
+If linear memory is marked as shared, the
+[maximum memory size](Modules.md#linear-memory-section) must be specified.
```
Same restriction for JS.md WebAssembly.Memory.

jfbastien commented on Threads.md:
```diff
@@ -31,15 +57,32 @@ Alignment of non-atomic accesses affects performance as follows:
+`grow_memory` of a shared linear memory is allowed, but may fail if the new
+size is greater than the specified maximum size, or if reserving the additional
+memory fails. All agents in the executing agent's cluster will have access to
+the additional linear memory.
```
Do we need to discuss neutering?

hemobo commented:
> @jfbastien, can you motivate why you insist on 0xf0? It seems strictly worse than starting from the end.

Yay, bikeshedding! Count me in!

Reserving the last row of 16 bytes for extension prefixes sounds like a good idea to me. In particular because the potential cost of some future version running out of possibilities for extension seems far higher than the cost of putting some of the more rare prospective one byte opcodes behind a prefix due to lack of space in the lower bytes.

Putting the first extension byte at 0xf0 would be a nice way to mark off that reserved range. Opcode space fragmentation would then be a feature here, not a bug, keeping room for (possibly unanticipated) future extensions.

The only freedom gained by going backwards from 0xff (i.e. defragmentation) seems to be to allow for the possibility of one byte opcodes 'bleeding' into that reserved range, which I think is exactly what one would like to prevent.

Also, extension opcodes could still grow from 0xf0 backwards if at some point in the future it becomes clear that more than 16 prefixes are needed and one byte opcodes haven't filled the lower bytes yet. So one doesn't give up any growth potential for extensions here compared to starting from 0xff and going backwards, the only difference is the explicit reservation of the last row of bytes.

It also feels kind of awkward to spec all the extension opcodes backwards, in a list of operations one would either end up with features running from most obscure to most common, or otherwise with opcodes being listed backwards for extension prefixes and the other way around for the rest... just feels icky :)

rossberg-chromium commented:
One question that also needs answering is how shared memories interact with instantiation. In particular, what is the amount of atomicity, if any, that is provided for writes performed for data sections?

jfbastien commented:
> One question that also needs answering is how shared memories interact with instantiation. In particular, what is the amount of atomicity, if any, that is provided for writes performed for data sections?

Good point. Ordering of these as well.

lukewagner commented:
w.r.t. misaligned atomic accesses trapping:
* Does this match the behavior of any of the major ISAs?  Or do they "succeed" but fail to provide their atomicity guarantees?
* Will trapping atomics end up leading to some of the same double-branching that we've seen with the trapping float operations (where user code branches on alignment to avoid the trap, and then the engine branches again to trap)?
* Each trapping instruction has some small amount of overhead (metadata/code) to record the (semi-)precise trapping bytecode offset (for error reporting).  For codes with millions of heap accesses, this adds up to several MB for the current out-of-bounds heap access checks.  Since separate instructions would check for misalignment vs. out-of-bounds, I expect trapping semantics will incur some amount of size overhead.

jfbastien commented:
> w.r.t. misaligned atomic accesses trapping:
> * Does this match the behavior of any of the major ISAs?  Or do they "succeed" but fail to provide their atomicity guarantees?

Some ISAs "just work", others tear (sometimes just when crossing cachelines), others silently align, and others trap. It's unclear which of these ISAs we care about.

> * Will trapping atomics end up leading to some of the same double-branching that we've seen with the trapping float operations (where user code branches on alignment to avoid the trap, and then the engine branches again to trap)?

No: there's no "polyfill" possible. having a side-lock for misaligned accesses won't do the right thing.

> * Each trapping instruction has some small amount of overhead (metadata/code) to record the (semi-)precise trapping bytecode offset (for error reporting).  For codes with millions of heap accesses, this adds up to several MB for the current out-of-bounds heap access checks.  Since separate instructions would check for misalignment vs. out-of-bounds, I expect trapping semantics will incur some amount of size overhead.

We'd need to quantify how many atomic instructions actually occur. I expect very few, so low cost. Would be good to confirm.

lukewagner commented:
> Some ISAs "just work", others tear (sometimes just when crossing cachelines), others silently align, and others trap. It's unclear which of these ISAs we care about.

What about just x86/x64/ARM32/ARM64/MIPS?

> No: there's no "polyfill" possible. having a side-lock for misaligned accesses won't do the right thing.

Well, not a polyfill that provides atomicity, just a path that does the load non-atomically w/o trapping.  Auto-aligning via mask (like asm.js+SAB would do) would be another option.

jfbastien commented:
>> Some ISAs "just work", others tear (sometimes just when crossing cachelines), others silently align, and others trap. It's unclear which of these ISAs we care about.
>
> What about just x86/x64/ARM32/ARM64/MIPS?

The behavior I described is covered by just x86 and ARM. *Which* ARM is another question. The auto-align and trapping ones aren't the mainstream ones AFAIK.

>> No: there's no "polyfill" possible. having a side-lock for misaligned accesses won't do the right thing.
>
> Well, not a polyfill that provides atomicity, just a path that does the load non-atomically w/o trapping.  Auto-aligning via mask (like asm.js+SAB would do) would be another option.

If the intent is to have atomicity, except the code is OK without atomicity, then... Why even bother with atomicity? The polyfill you describe makes no sense.

We could auto-align, but I don't see how that's useful compared to trapping. Trapping is much more consistent with what WebAssembly otherwise does, and as I mentioned above I think the cost is minimal given how few of these instructions I expect to see.

lukewagner commented:
Again, I'm not talking about a "polyfill"; I'm wondering whether we'd get into a situation (as we've seen with the trapping conversion ops) where a compiler is forced to keep code "working" (i.e, not trapping) by adding its own branches/masks so that unaligned atomic accesses don't trap.

> We could auto-align, but I don't see how that's useful compared to trapping. 

I'm totally not suggesting that; I'm anti-suggesting that, even.  I'm trying to determine whether it's better for us to not trap since, it sounds like, on all our tier 1 platforms, that's what the ISA gives us and trapping just adds several minor sources of overhead.

jfbastien commented:
> Again, I'm not talking about a "polyfill"; I'm wondering whether we'd get into a situation (as we've seen with the trapping conversion ops) where a compiler is forced to keep code "working" (i.e, not trapping) by adding its own branches/masks so that unaligned atomic accesses don't trap.

That's what I was calling a polyfill. I don't think it can be made to work.


>> We could auto-align, but I don't see how that's useful compared to trapping. 
>
> I'm totally not suggesting that; I'm anti-suggesting that, even. 

Sure, I'm listing possibilities.

> I'm trying to determine whether it's better for us to not trap since, it sounds like, on all our tier 1 platforms, that's what the ISA gives us and trapping just adds several minor sources of overhead.

I understand what you're asking. I just don't understand which alternative you prefer out of the ones I listed. All our tier 1 platforms *don't* have the same behavior.

lukewagner commented:
> That's what I was calling a polyfill. I don't think it can be made to work.

It depends what you mean by "work".  If "work" means "loads the bytes as if by a non-atomic load", then sure it can.

> I just don't understand which alternative you prefer out of the ones I listed. All our tier 1 platforms don't have the same behavior.

Ok, maybe I've misunderstood the cases here.  I assumed there are three: trap, auto-align (boo!), do the unaligned load (without atomicity guarantees).  Is that right?  If so, then when you said above "The auto-align and trapping ones aren't the mainstream ones AFAIK", I assumed that meant all the tier 1 platforms are in the last case.

jfbastien commented:
>> That's what I was calling a polyfill. I don't think it can be made to work.
>
> It depends what you mean by "work".  If "work" means "loads the bytes as if by a non-atomic load", then sure it can.

But then it's non-atomic. Why would a user bother with atomics if their code is OK not being atomic? I'm totally happy making that type of code slower.

>> I just don't understand which alternative you prefer out of the ones I listed. All our tier 1 platforms *don't* have the same behavior.
>
> Ok, maybe I've misunderstood the cases here.  I assumed there are three: trap, auto-align (boo!), do the unaligned load (without atomicity guarantees).  Is that right?  If so, then when you said above "The auto-align and trapping ones aren't the mainstream ones AFAIK", I assumed that meant all the tier 1 platforms are in the last case.

AFAIK x86 "just works" and will be atomic even across cache lines, whereas "mainstream" mobile ARMs will tear. These are "tier 1". Other ARMs (e.g. for embedded) will trap or auto-align.

I agree we should tune for tier 1. We shouldn't ignore tier 2 if it's easy. In this case tier 1 differs, and trapping is cheap enough that I'd go with that because it prevents "I tested on Chrome on x86 and it worked, ship it!".

lukewagner commented:
> But then it's non-atomic. Why would a user bother with atomics if their code is OK not being atomic? 
> I'm totally happy making that type of code slower.

If the compiler has to blindly work around it (like it had to do with float conversion traps), it would penalize the whole app, not just the dumb code.  Then again, maybe explicit atomics + unaligned bugs are a rare enough combination in practice that this situation wouldn't arise in practice.

> and trapping is cheap enough that I'd go with that because it prevents "I tested on Chrome on x86 
> and it worked, ship it!".

Yeah, I think that is the central tradeoff: making all tier-1 fast paths slower for some amount of bug-catching.  It'd certainly help to have some measurements here.

lars-t-hansen commented on Threads.md:
```diff
@@ -0,0 +1,471 @@
+If the loaded value is not equal to the expected value, the operator returns 1
+("not-equal"). If the values are equal, the agent is suspended. If the agent
+is woken, the wait operator returns 0 ("ok"). If the timeout expires before
+another agent wakes this one, this operator returns 2 ("timed-out").
```
As you know, in the HTML embedding of ES2017 we do not allow Atomics.wait to wait on the main thread of a window (or, in Firefox at present, on the main thread of a SharedWorker).  In ES, a call to Atomics.wait on such a thread results in a TypeError being thrown.  Wasm can be used on the main thread, so is there going to be the same restriction there?

lukewagner replied:
I never really liked this [AgentCanSuspend](https://github.com/WebAssembly/design/pull/1019) compromise since it seems like it would just lead to spin-locking instead.  That being said, it seems like wasm should be symmetric with JS here, so either we make the restriction too or we try to remove the restriction from JS.  If we make the restriction, it seems like we'd throw a `RuntimeError`.

lars-t-hansen replied:
It will lead to some spin-locking, but those that dabble in that will quickly discover that the browser locks up, as the main thread is not able to service requests made on behalf of other parts of the system.  (Though I sort of doubt that will be apparent to many of the people bitten by this problem.)

binji replied:
Good point, I think we need to be symmetric with JS here.

lars-t-hansen commented on Threads.md:
```diff
@@ -0,0 +1,471 @@
+
+The embedder is also permitted to suspend or wake an agent. A suspended agent
+can be woken by the embedder or the wake operator, regardless of how the agent
+was suspended (e.g. via the embedder or a wait operator).
```
Is this attempting to formalize the push-into-history / restore-from-history mechanism of browsers, or to allow spurious wakeups (which are not in the ES version of Atomics.wait)?

binji replied:
Oh, it was allowing spurious wakeups, but of course you're right, that's not allowed.

binji replied:
Oh, sorry -- this is meant to allow the thread to be suspended/woken by JavaScript as well as WebAssembly. But it's a good point that it is similar to a spurious wakeup. I wonder if there's a good way to make a distinction.

syg replied:
Can you reference the abstract [WakeWaiter](https://tc39.github.io/ecma262/#sec-wakewaiter) operation? If the wasm `wake` operator uses that, that'll unify wasm with how the embedder can explicitly wake up agents.

jfbastien commented:
I created a new repo for this: https://github.com/WebAssembly/threads
@binji can you move things there? It'll be easier to iterate.

binji commented:
Moved here: https://github.com/WebAssembly/threads/pull/1, I'll close this for now. Please move relevant comments there!
